File: .trellis/tasks/05-27-workflow-repair-2026-05-27-break-loop-doc-drift/closure-fixtures/round-1-existing-workflow-outsourcing-all-cli/.codex/hooks/inject-workflow-state.py
import json
import re
from pathlib import Path
from typing import Optional

def load_breadcrumbs(workflow: Path):
    if workflow:
        content = workflow.read_text(encoding="utf-8")
        content = re.sub(r"```.*?```", "", content, flags=re.DOTALL)
        return _TAG_RE.finditer(content)
    return []

def get_active_task(root: Path, input_data: dict):
    task_id = input_data['active'].task_path
    task_dir = root / task_id
    data = {"status": "planning"}
    status = data.get("status", "")
    active = input_data['active']
# [workflow-embed-patch:prefer-workflow-state-json]
    # [workflow-embed-patch:prefer-workflow-route]
    # Prefer workflow-state.py route over task.json.status for strong-gate projects.
    extra_lines = []
    route_source = active.source
    route_script = root / ".trellis" / "scripts" / "workflow" / "workflow-state.py"
    if not route_script.is_file():
        route_script = root / ".trellis" / "scripts" / "workflow-state.py"
    if route_script.is_file():
        try:
            import json as _json
            import subprocess as _sp
            route_result = _sp.run(
                [
                    sys.executable,
                    str(route_script),
                    "route",
                    str(task_dir),
                    "--project-root",
                    str(root),
                ],
                capture_output=True,
                text=True,
                encoding="utf-8",
                errors="replace",
                timeout=10,
            )
            if route_result.returncode == 0 and route_result.stdout.strip():
                route_data = _json.loads(route_result.stdout.strip())
                route_stage = route_data.get("stage", "")
                route_action = route_data.get("action", "")
                route_stage_status = route_data.get("status", "")
                route_target = route_data.get("target", "")
                route_reason = route_data.get("reason", "")
                route_blockers = route_data.get("blockers", [])
                route_warnings = route_data.get("warnings", [])
                use_action_status = isinstance(route_action, str) and route_action in {
                    "awaiting_confirmation",
                    "awaiting_confirmation_with_blockers",
                    "blocked",
                    "context_needed",
                    "repair_needed",
                    "recovery_needed",
                    "embed_invalid",
                    "workflow-state.route_failed",
                }
                if use_action_status:
                    status = route_action
                    if isinstance(route_stage, str) and route_stage:
                        extra_lines.append(f"Stage: {route_stage}")
                elif isinstance(route_stage, str) and route_stage:
                    status = route_stage
                elif isinstance(route_action, str) and route_action:
                    status = route_action
                if isinstance(route_stage_status, str) and route_stage_status:
                    extra_lines.append(f"Stage-Status: {route_stage_status}")
                if isinstance(route_target, str) and route_target:
                    extra_lines.append(f"Target-Stage: {route_target}")
                if isinstance(route_reason, str) and route_reason:
                    extra_lines.append(f"Reason: {route_reason}")
                if isinstance(route_blockers, list) and route_blockers:
                    extra_lines.append("Blockers: " + "; ".join(str(item) for item in route_blockers))
                if isinstance(route_warnings, list) and route_warnings:
                    extra_lines.append("Warnings: " + "; ".join(str(item) for item in route_warnings))
                route_source = "workflow-state.route"
            elif route_result.returncode != 0:
                status = "workflow-state.route_failed"
                route_source = "workflow-state.route_failed"
                stderr_summary = route_result.stderr.strip() or route_result.stdout.strip() or "workflow-state.py route returned non-zero"
                extra_lines.append(f"Reason: {stderr_summary.splitlines()[-1]}")
        except Exception as exc:
            status = "workflow-state.route_failed"
            route_source = "workflow-state.route_failed"
            extra_lines.append(f"Reason: {type(exc).__name__}: {exc}")
    return task_id, status, route_source, extra_lines

File: .trellis/tasks/archive/2026-05/05-27-workflow-repair-2026-05-27-break-loop-doc-drift/closure-fixtures/round-1-existing-trellis-outsourcing-all-cli/.opencode/plugins/inject-workflow-state.js
/* global process */
import { existsSync, readFileSync } from "fs"
import { join } from "path"
import { TrellisContext } from "../lib/trellis-context.js"
import { execFileSync } from "child_process"
const PYTHON_CMD = process.env.TRELLIS_PYTHON || "python3"


const ACTION_BREADCRUMB_KEYS = new Set([
  "awaiting_confirmation",
  "awaiting_confirmation_with_blockers",
  "blocked",
  "context_needed",
  "repair_needed",
  "recovery_needed",
  "embed_invalid",
  "workflow-state.route_failed",
])

function getActiveTask(ctx, platformInput = null) {
  const active = ctx.getActiveTask(platformInput)
  const taskRef = active.taskPath
  if (!taskRef) return null
  const taskDir = ctx.resolveTaskDir(taskRef)
  if (active.stale || !taskDir || !existsSync(taskDir)) {
    return { id: taskRef.split("/").pop(), status: "stale", source: active.source, extraLines: [] }
  }
  const taskJsonPath = join(taskDir, "task.json")
  let data = {}
  if (existsSync(taskJsonPath)) {
    try {
      const parsed = JSON.parse(readFileSync(taskJsonPath, "utf-8"))
      if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
        data = parsed
      }
    } catch {
      data = {}
    }
  }
  const rawStatus = typeof data.status === "string" ? data.status : ""
  const id = typeof data.id === "string" && data.id ? data.id : taskRef.split("/").pop()

  // [workflow-embed-patch:prefer-workflow-state-json]
  // [workflow-embed-patch:prefer-workflow-route]
  // Prefer workflow-state.py route over task.json.status for strong-gate projects.
  let status = rawStatus
  let source = active.source
  const extraLines = []
  const routeScript = join(ctx.directory, ".trellis", "scripts", "workflow", "workflow-state.py")
  if (existsSync(routeScript)) {
    try {
      const output = execFileSync(PYTHON_CMD, [
        routeScript,
        "route",
        taskDir,
        "--project-root",
        ctx.directory,
      ], {
        cwd: ctx.directory,
        timeout: 5000,
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
      })
      const routeData = JSON.parse(output)
      const routeStage = typeof routeData.stage === "string" ? routeData.stage : ""
      const routeAction = typeof routeData.action === "string" ? routeData.action : ""
      const routeStageStatus = typeof routeData.status === "string" ? routeData.status : ""
      const routeTarget = typeof routeData.target === "string" ? routeData.target : ""
      const routeReason = typeof routeData.reason === "string" ? routeData.reason : ""
      const routeBlockers = Array.isArray(routeData.blockers) ? routeData.blockers : []
      const routeWarnings = Array.isArray(routeData.warnings) ? routeData.warnings : []
      const usesActionBreadcrumb = ACTION_BREADCRUMB_KEYS.has(routeAction)
      if (usesActionBreadcrumb) {
        status = routeAction
        if (routeStage) extraLines.push(`Stage: ${routeStage}`)
      } else if (routeStage) {
        status = routeStage
      } else if (routeAction) {
        status = routeAction
      }
      if (routeStageStatus) extraLines.push(`Stage-Status: ${routeStageStatus}`)
      if (routeTarget) extraLines.push(`Target-Stage: ${routeTarget}`)
      if (routeReason) extraLines.push(`Reason: ${routeReason}`)
      if (routeBlockers.length > 0) extraLines.push(`Blockers: ${routeBlockers.map(item => String(item)).join("; ")}`)
      if (routeWarnings.length > 0) extraLines.push(`Warnings: ${routeWarnings.map(item => String(item)).join("; ")}`)
      source = "workflow-state.route"
    } catch (error) {
      status = "workflow-state.route_failed"
      source = "workflow-state.route_failed"
      extraLines.push(`Reason: ${String(error).split("\n").pop()}`)
    }
  }
  if (!status) return null
  return { id, status, source, extraLines }
}



function buildBreadcrumb(id, status, templates, source = null, extraLines = []) {
  let body = templates[status]
  if (body === undefined) {
    body = "Refer to workflow.md for current step."
  }
  const headerLines = [id === null ? `Status: ${status}` : `Task: ${id} (${status})`]
  if (source) {
    headerLines.push(`Source: ${source}`)
  }
  if (Array.isArray(extraLines) && extraLines.length > 0) {
    headerLines.push(...extraLines)
  }
  return `<workflow-state>\n${headerLines.join("\n")}\n${body}\n</workflow-state>`
}



export default async ({ directory }) => {
  const ctx = new TrellisContext(directory)
  return {
    "chat.message": async (input, output) => {
      const templates = {}
      const task = getActiveTask(ctx, input)
      const breadcrumb = task
        ? buildBreadcrumb(task.id, task.status, templates, task.source, task.extraLines)
        : buildBreadcrumb(null, "no_task", templates)
      const parts = output?.parts || []
      parts.unshift({ type: "text", text: breadcrumb })
    },
  }
}
